wrapper/src/main/java/software/amazon/jdbc/hostlistprovider/RdsMultiAzDbClusterListProvider.java

```diff
@@ -0,0 +1,197 @@
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.hostlistprovider;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLSyntaxErrorException;
+import java.sql.Statement;
+import java.sql.Timestamp;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Properties;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+import software.amazon.jdbc.HostListProviderService;
+import software.amazon.jdbc.HostRole;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.hostavailability.HostAvailability;
+import software.amazon.jdbc.util.Messages;
+
+public class RdsMultiAzDbClusterListProvider extends RdsHostListProvider {
+  private final String fetchWriterNodeQuery;
+  private final String fetchWriterNodeQueryHeader;
+  static final Logger LOGGER = Logger.getLogger(RdsMultiAzDbClusterListProvider.class.getName());
+
+  public RdsMultiAzDbClusterListProvider(
+      final Properties properties,
+      final String originalUrl,
+      final HostListProviderService hostListProviderService,
+      final String topologyQuery,
+      final String nodeIdQuery,
+      final String isReaderQuery,
+      final String fetchWriterNodeQuery,
+      final String fetchWriterNodeQueryHeader
+  ) {
+    super(properties,
+        originalUrl,
+        hostListProviderService,
+        topologyQuery,
+        nodeIdQuery,
+        isReaderQuery);
+    this.fetchWriterNodeQuery = fetchWriterNodeQuery;
+    this.fetchWriterNodeQueryHeader = fetchWriterNodeQueryHeader;
+  }
+
+  /**
+   * Obtain a cluster topology from database.
+   *
+   * @param conn A connection to database to fetch the latest topology.
+   * @return a list of {@link HostSpec} objects representing the topology
+   * @throws SQLException if errors occurred while retrieving the topology.
+   */
+  protected List<HostSpec> queryForTopology(final Connection conn) throws SQLException {
+
+    int networkTimeout = -1;
+    try {
+      networkTimeout = conn.getNetworkTimeout();
+      // The topology query is not monitored by the EFM plugin, so it needs a socket timeout
+      if (networkTimeout == 0) {
+        conn.setNetworkTimeout(networkTimeoutExecutor, defaultTopologyQueryTimeoutMs);
+      }
+    } catch (SQLException e) {
+      LOGGER.warning(() -> Messages.get("RdsHostListProvider.errorGettingNetworkTimeout",
+          new Object[] {e.getMessage()}));
+    }
+
+    try {
+      final Statement stmt = conn.createStatement();
+      String writerNodeId = processWriterNodeId(stmt.executeQuery(this.fetchWriterNodeQuery));
+      if (writerNodeId.isEmpty()) {
+        final ResultSet nodeIdResultSet = stmt.executeQuery(this.nodeIdQuery);
+        while (nodeIdResultSet.next()) {
+          writerNodeId = nodeIdResultSet.getString(1);
+        }
+      }
```

Comment:
This block is unclear. If writer is empty, it fetches topology and consider the last node in this topology as a writer. Is that correct?

Reply:
This part is still being discussed and subject to change. For current implementation:
* If `SHOW REPLICA STATUS` returns a Source node ID, it implies that the current node is read replica.
* If `SHOW REPLICA STATUS` returns nothing, it means that the current node is not replicating from other nodes, implies that the current node is writer.